doc/adr/0004-common-pab-api.rst

```diff
@@ -0,0 +1,44 @@
+.. _common_pab_api:
+
+ADR 4: Common PAB API
+=====================
+
+Date: 2022-07-12
+
+Authors
+---------
+
+Gergely Szabo <gergely@mlabs.city>
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+There are multiple implementations of a Plutus backend, and also other tools related to Plutus smart contracts. Some of them are using the same contract interface as the official implementation, but some of them use a different interface. However, as the ecosystem evolves, it would be beneficial to create a well defined standard, that other off-chain tools can use as a reference, or as an interface to implement.
+
+Currently, as we are getting close to the Vasil hardfork, testing tools and Plutus backend tools are at a hurry to update their dependencies and get to a Vasil compliant/compatible state. However, tools that are depending on `plutus-apps` are blocked by the PAB development. This initiative was born out of this context, but could solve other problems as well.
+
+Benefits
+--------
+
+The idea proposed in this document could speed up the development of off-chain tools, by loosening up some of tightly coupled dependencies, so these projects can move more freely. This would also mean that the cost of the interface update would be reduced, so we could see more features added to the standard, and the PAB API following the capabilities of Cardano more closely. As an added benefit, community involvement with the API could also greatly improve.
+
+A standard API for all Plutus contacts would help keeping the ecosystem on the same track with their implementation. As more and more off-chain tools implement the same contract interface in the future, it will be relatively easy to switch between different PAB implementations, or to use multiple of these tools at the same time without a need for serious code rewrites.
+
+Proposal
+--------
+
+The PAB Contract (defined in `plutus-apps`) is using the `freer-simple` effect system to define all the contract effects. This already allows us to separate the interface from the implementation, or to have multiple implementations for one interface.
+
+Our proposal, is to further separate these two components: create a separate repository with these contract effects and types, but without any implementation or any unnecessary dependencies. By moving the contracts out of the plutus-apps monorepository, any tool could update to newer version to their discretion. Without many dependencies, many tools could utilize the contract API without having to depend on the whole plutus-apps monorepo.
```

Comment:
I think having the contract effects and types in a separate package should be enough right? I don't see the need for a separate repository.

Reply:
The disadvantage of this is that cabal will have to clone the whole repo to use the package (https://github.com/haskell/cabal/issues/7264).

Reply:
That's a downside that seems pretty minor to us not having a monorepo anymore.

Reply:
Another downside is that PAB will not be able to use a different version of the interface. Tools, that only use the interface would have to wait for the PAB implementation. And the reality is that the PAB implementation is lagging behind the hardfork, and we're unsure when we will be able to use Vasil features (this is the main reason for this proposal).

Reply:
Ideally, the interface would be updated well before a testnet hardfork, so library developers can start building on it.

Reply:
Right, we could do that with a separate branch, like we do now with the `next-node` branch, but that's a bit of a kludge.

Reply:
It'd also allow this specific "interface-only" package not depend on a great amount of packages that may also need to be updated for it to work. A separate repo will ensure the interface package is truly minimal, easy to depend on, and implement an interpreter for. Of course, it'll also make this interface-only repo much easier to maintain.